<!DOCTYPE html>
<html lang="en">
<head>

  <meta charset="utf-8" />

  
  <title>Chenhao&#39;s blog posts</title>

  
  




  
  <meta name="author" content="Chenhao Li (李陈浩)" />
  <meta name="description" content="" />

  
  <meta name="twitter:card" content="summary" />
  <meta name="twitter:site" content="@li_chenhao" />
  <meta name="twitter:title" content="Chenhao&#39;s blog posts" />
  <meta name="twitter:description" content="" />
  <meta name="twitter:image" content="/img/avatar.jpg" />




<meta name="generator" content="Hugo 0.30.2" />


<link rel="canonical" href="/" />
<link rel="alternative" href="/index.xml" title="Chenhao&#39;s Blog" type="application/atom+xml" />


<meta name="renderer" content="webkit" />
<meta name="viewport" content="width=device-width,initial-scale=1" />
<meta name="format-detection" content="telephone=no,email=no,adress=no" />
<meta http-equiv="Cache-Control" content="no-transform" />


<meta name="robots" content="index,follow" />
<meta name="referrer" content="origin-when-cross-origin" />







<meta name="apple-mobile-web-app-capable" content="yes" />
<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />
<meta name="apple-mobile-web-app-title" content="Chenhao&#39;s Blog" />
<meta name="msapplication-tooltip" content="Chenhao&#39;s Blog" />
<meta name='msapplication-navbutton-color' content="#5fbf5e" />
<meta name="msapplication-TileColor" content="#5fbf5e" />
<meta name="msapplication-TileImage" content="/img/tile-image-windows.png" />
<link rel="icon" href="/img/favicon.ico" />
<link rel="icon" type="image/png" sizes="16x16" href="/img/favicon-16x16.png" />
<link rel="icon" type="image/png" sizes="32x32" href="/img/favicon-32x32.png" />
<link rel="icon" sizes="192x192" href="/img/touch-icon-android.png" />
<link rel="apple-touch-icon" href="/img/touch-icon-apple.png" />
<link rel="mask-icon" href="/img/safari-pinned-tab.svg" color="#5fbf5e" />



<link rel="stylesheet" href="/css/bundle.css" />


  
  <!--[if lt IE 9]>
    <script src="//cdn.bootcss.com/html5shiv/3.7.3/html5shiv.min.js"></script>
    <script src="//cdn.bootcss.com/respond.js/1.4.2/respond.min.js"></script>
    <script src="//cdn.bootcss.com/video.js/6.2.8/ie8/videojs-ie8.min.js"></script>
  <![endif]-->

<!--[if lte IE 11]>
    <script src="//cdn.bootcss.com/classlist/1.1.20170427/classList.min.js"></script>
  <![endif]-->


<script src="//cdn.bootcss.com/object-fit-images/3.2.3/ofi.min.js"></script>


<script src="//cdn.bootcss.com/smooth-scroll/12.1.4/js/smooth-scroll.polyfills.min.js"></script>


</head>
  <body>
    
    <div class="suspension">
      <a title="Go to top" class="to-top is-hide"><span class="icon icon-up"></span></a>
      
    </div>
    
    
  <header class="site-header">
  <img class="avatar" src="/img/avatar.jpg" alt="Avatar">
  
  <h1 class="title">Chenhao&#39;s Blog</h1>
  
  <p class="subtitle">~ Think as a scientist &amp; Work as an engineer ~</p>
  <button class="menu-toggle" type="button">
    <span class="icon icon-menu"></span>
  </button>
  <nav class="site-menu collapsed">
    <h2 class="offscreen">Main Menu</h2>
    <ul class="menu-list">
      
      
      
      
        <li class="menu-item  is-active"><a href="/">Home</a></li>
      
        <li class="menu-item "><a href="https://github.com/lch14forever">Works</a></li>
      
        <li class="menu-item "><a href="/links/">Links</a></li>
      
        <li class="menu-item "><a href="/about/">About</a></li>
      
    </ul>
  </nav>
  <nav class="social-menu collapsed">
    <h2 class="offscreen">Social Networks</h2>
    <ul class="social-list">

      
      <li class="social-item">
        <a href="mailto:lichenhao.sg@gmail.com" title="Email"><span class="icon icon-email"></span></a>
      </li>

      
      <li class="social-item">
        <a href="//github.com/lch14forever" title="GitHub"><span class="icon icon-github"></span></a>
      </li>

      <li class="social-item">
        <a href="//twitter.com/li_chenhao" title="Twitter"><span class="icon icon-twitter"></span></a>
      </li>

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      <li class="social-item">
        <a href="//www.linkedin.com/in/chenhao-li-1b695b99" title="Linkedin"><span class="icon icon-linkedin"></span></a>
      </li>

      

      

      

      <li class="social-item">
        <a href="/index.xml"><span class="icon icon-rss" title="RSS"></span></a>
      </li>

    </ul>
  </nav>
</header>

  <section class="main post-list">
    <header class="list-header offscreen">
      <h2 class="list-label">All Posts</h2>
    </header>
    
    
      
        <article class="post-entry">
  <header class="post-header">
    <h3 class="post-title"><a href="/post/2017-11-18-r-absolute/" class="post-link">有了绝对定量就完了？</a></h3>
    <p class="post-meta">@Chenhao Li · Nov 28, 2017 · 5 min read</p>
  </header>
  
  <p class="post-summary">微生物组研究走向绝对定量 Jeroen Raes研究组上周在Nature发表文章，使用Flow cytometry估算粪便中的微生物细胞数量，配合16S测序算出的相对丰度，估算出OTU的绝对数量（absolute abundance）。文章很多结论都印证了直接应用相对丰度进行分析时，我们所看到的很多现象是成分数据性质的假象（可以想象，由于相对丰度在每一个样本中相加为1或100，一个OTU相对丰度增加必将引起其他OTU相对丰度减少，所谓的compositional bias）。特别针对于计算两个OTU的相关系数，当OTU分布不均匀时，很容易看到负相关的OTU–而这仅仅是因为它们受到相加为常数的限制而已。另一个典型例子就是主成分分析（PCA），PCA意在保持欧式距离不变的情况下对数据进行变换，但是相对丰度其实不在欧式空间中（可以参考:J. Aitchison, The Statistical Analysis of Compositional Data, 1986.），这就是为什么在微生物组的研究中更多采用生态学的距离（如，Bray-Curtis distance）来计算\(\beta\) -diversity，然后进行基于距离矩阵的分析（PCoA）。
似乎文章的take home message很简单了，微生物组的研究，我们应该使用类似的方法进行绝对定量。可是仔细想一想，文章指出粪便微生物总量的个体差异可以达到10倍之多，这样大的差异，如果某种微生物在个体之间差异很小，转换成绝对数量之后，个体间的差异将受制于微生物总量。
 使用Flow cytometry测定的肠道微生物总量的variation有多大？ 下载Nature文章中的Supplementary Table，这里并不不需要购买文章阅读权限，其中表6是Flow cytometry的数据。
加载R包，ggplot2主题 library(tidyverse) ## ── Attaching packages ──────────────────────────────────────────────────────── tidyverse 1.2.1 ── ## ✔ ggplot2 2.2.1 ✔ purrr 0.2.4 ## ✔ tibble 1.3.4 ✔ dplyr 0.7.4 ## ✔ tidyr 0.7.2 ✔ stringr 1.2.0 ## ✔ readr 1.1.1 ✔ forcats 0.2.0 ## ── Conflicts ─────────────────────────────────────────────────────────── tidyverse_conflicts() ── ## ✖ dplyr::filter() masks stats::filter() ## ✖ dplyr::lag() masks stats::lag() library(data.</p>
  <footer class="post-footer">
    <a class="read-more" href="/post/2017-11-18-r-absolute/">Read More →</a>
  </footer>
</article>

      
        <article class="post-entry">
  <header class="post-header">
    <h3 class="post-title"><a href="/post/2017-11-15-r-glv/" class="post-link">Generalized Lokta-Volterra model</a></h3>
    <p class="post-meta">@Chenhao Li · Nov 15, 2017 · 1 min read</p>
  </header>
  
  <p class="post-summary">The model \(\frac{dx_i}{dt}= ax_i&#43;\sum_{j=1}^pb_{ij}x_ix_j\)
## library library(&#39;deSolve&#39;) ## GLV function lvm &lt;- function(t, x, params){ with(as.list(params, c(x)), { dx &lt;- alpha * x &#43; x * (beta %*% x) list(dx) }) } ## numerical integration n.integrate &lt;- function(time, init.x, model, params){ as.data.frame(ode(init.x, time, model, params)) } ## normalizations TSS &lt;- function(x){ apply(x, 1, function(x)x/sum(x)) }  GLV with 3 species alpha &lt;- c(0.1, 0.2, 0.3) beta &lt;- t(matrix(c(-0.5, -0.3 , 0, 0.</p>
  <footer class="post-footer">
    <a class="read-more" href="/post/2017-11-15-r-glv/">Read More →</a>
  </footer>
</article>

      
        <article class="post-entry">
  <header class="post-header">
    <h3 class="post-title"><a href="/post/2017-11-07-r-curatedmetagenomicdata_retrievedata/" class="post-link">从curatedMetagenoimcData提取健康样本的微生物组</a></h3>
    <p class="post-meta">@Chenhao Li (李陈浩) · Nov 7, 2017 · 2 min read</p>
  </header>
  
  <p class="post-summary">简介 有人问我，这个curatedMetagenomicData有什么作用？对于我们这些research parasite(出处请见NEJM社论)来说，这样庞大的数据集当然是用来挖信息或者是测试算法了：
作为validation cohort来验证已发现的结论。如我们的这篇文章都使用了2014年Oh et al发表于Nature的数据来佐证我们发现的Staphylococcus特异性突变。
 用来测试新方法，发现新的生物问题。如这篇文章中我们也用了Oh et al的数据来观测Malessezia在人体皮肤的分布。
 因为数据中有大量健康人的微生物组（不同研究中的control），我们也可以用它们补充我们的对照组（当然在机器学习中要注意数据不平衡问题）。
  当然最基础的操作，就是要从数据库中提取数据。
 演示 我们从curatedMetagenomicData中提取健康人的粪便及皮肤微生物组。
加载R package suppressMessages(library(curatedMetagenomicData)) ## Warning: package &#39;dplyr&#39; was built under R version 3.4.2 ## Warning: package &#39;BiocGenerics&#39; was built under R version 3.4.1 ## Warning: package &#39;ExperimentHub&#39; was built under R version 3.4.1 ## Warning: package &#39;AnnotationHub&#39; was built under R version 3.4.2 suppressMessages(library(foreach))  下载并合并所有粪便、皮肤数据（下载数据缓存于~/.ExperimentHub中，第二次以后会直接读取缓存） ## download all stool samples stoolData &lt;- curatedMetagenomicData(&#34;*.</p>
  <footer class="post-footer">
    <a class="read-more" href="/post/2017-11-07-r-curatedmetagenomicdata_retrievedata/">Read More →</a>
  </footer>
</article>

      
        <article class="post-entry">
  <header class="post-header">
    <h3 class="post-title"><a href="/post/2017-11-04-nanopore-nanoporedataanalysis/" class="post-link">A collection of useful tools (Keep updating)</a></h3>
    <p class="post-meta">@Chenhao Li · Nov 4, 2017 · 1 min read</p>
  </header>
  
  <p class="post-summary">General bioinformatics  Applied genomics course from Michael Schatz https://github.com/schatzlab/appliedgenomics Data science from Rafael A. Irizarry (built with R bookdown) https://rafalab.github.io/dsbook/   Nanopore seqeuncing data analysis  Assembly best practice QC Adaptor removal with Porechop (Minimap2 &#43; Miniasm ) or Canu &#43; Racon &#43; Nanopolish    Metagenomics  Coming soon …   Single cell analysis  Tool collection for SC data https://github.com/seandavi/awesome-single-cell   </p>
  <footer class="post-footer">
    <a class="read-more" href="/post/2017-11-04-nanopore-nanoporedataanalysis/">Read More →</a>
  </footer>
</article>

      
        <article class="post-entry">
  <header class="post-header">
    <h3 class="post-title"><a href="/post/2017-11-02-r-sankeyplot2/" class="post-link">桑基图（Sankey diagram）II</a></h3>
    <p class="post-meta">@Chenhao Li (李陈浩) · Nov 2, 2017 · 3 min read</p>
  </header>
  
  <p class="post-summary">使用plotly 预处理 taxtab &lt;- read.table(&#39;SRS014476-Supragingival_plaque_profile.txt&#39;, sep=&#39;\t&#39;, stringsAsFactors=FALSE) second_last &lt;- function(x) ifelse(length(x)&gt;1, x[length(x)-1], NA) taxtab &lt;- taxtab[-grep(&#39;t__.*unclassified&#39;, taxtab$V1), ] ## remove strain level tax_split &lt;- strsplit(taxtab$V1, &#39;\\|&#39;) ## split into different taxonomy levels target &lt;- sapply(tax_split, tail, n=1) ## target node mapping &lt;- data.frame(id=0:(length(target)-1), row.names=target) target_id &lt;- mapping$id source &lt;- sapply(tax_split, second_last) ## source node source_id &lt;- mapping[source,] value &lt;- taxtab$V2 ## width of flow links &lt;- data.</p>
  <footer class="post-footer">
    <a class="read-more" href="/post/2017-11-02-r-sankeyplot2/">Read More →</a>
  </footer>
</article>

      
        <article class="post-entry">
  <header class="post-header">
    <h3 class="post-title"><a href="/post/2017-10-30-r-curatedmetagenomicdata/" class="post-link">超大宏基因组数据集CuratedMetagenomicData</a></h3>
    <p class="post-meta">@Chenhao Li (李陈浩) · Oct 30, 2017 · 6 min read</p>
  </header>
  
  <p class="post-summary">简介 CuratedMetagenomicData(https://waldronlab.github.io/curatedMetagenomicData/)的目标是用标准化的流程（MetaPhlan2、HUMAnN2）分析已发表的宏基因组数据并建立一个统一的数据集合。目前已经收录6000余个样本并在持续扩建当中。样本涵盖糖尿病、肥胖症、IBD等多种疾病，涉及皮肤、口腔、粪便等多处样本。
 完整帮助信息 https://bioconductor.org/packages/devel/data/experiment/vignettes/curatedMetagenomicData/inst/doc/curatedMetagenomicData.html
 安装 通过Bioconductor安装最新版本
## try http:// if https:// URLs are not supported source(&#34;https://bioconductor.org/biocLite.R&#34;) useDevel() biocLite(&#34;curatedMetagenomicData&#34;)  基本使用 调用
suppressPackageStartupMessages(library(curatedMetagenomicData)) ## Warning: package &#39;dplyr&#39; was built under R version 3.4.2 ## Warning: package &#39;BiocGenerics&#39; was built under R version 3.4.1 ## Warning: package &#39;ExperimentHub&#39; was built under R version 3.4.1 ## Warning: package &#39;AnnotationHub&#39; was built under R version 3.4.2 查看metadata
combined_metadata ## # A tibble: 6,058 x 80 ## dataset_name sampleID subjectID body_site antibiotics_current_use ## &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; ## 1 AsnicarF_2017 MV_FEI1_t1Q14 MV_FEI1 stool &lt;NA&gt; ## 2 AsnicarF_2017 MV_FEI2_t1Q14 MV_FEI2 stool &lt;NA&gt; ## 3 AsnicarF_2017 MV_FEI3_t1Q14 MV_FEI3 stool &lt;NA&gt; ## 4 AsnicarF_2017 MV_FEI4_t1Q14 MV_FEI4 stool &lt;NA&gt; ## 5 AsnicarF_2017 MV_FEI4_t2Q15 MV_FEI4 stool &lt;NA&gt; ## 6 AsnicarF_2017 MV_FEI5_t1Q14 MV_FEI5 stool &lt;NA&gt; ## 7 AsnicarF_2017 MV_FEI5_t2Q14 MV_FEI5 stool &lt;NA&gt; ## 8 AsnicarF_2017 MV_FEI5_t3Q15 MV_FEI5 stool &lt;NA&gt; ## 9 AsnicarF_2017 MV_FEM1_t1Q14 MV_FEM1 stool &lt;NA&gt; ## 10 AsnicarF_2017 MV_FEM2_t1Q14 MV_FEM2 stool &lt;NA&gt; ## # .</p>
  <footer class="post-footer">
    <a class="read-more" href="/post/2017-10-30-r-curatedmetagenomicdata/">Read More →</a>
  </footer>
</article>

      
        <article class="post-entry">
  <header class="post-header">
    <h3 class="post-title"><a href="/post/2017-10-29-r-sankeyplot/" class="post-link">桑基图（Sankey diagram）</a></h3>
    <p class="post-meta">@Chenhao Li (李陈浩) · Oct 29, 2017 · 1 min read</p>
  </header>
  
  <p class="post-summary">简介 桑基图（Sankey diagram）是用于表示能量或信息流动的一种可视化方式，应用于微生物组数据，可以清晰展示各个taxonomy level之间物种相对丰度的流动。从Domain到Species，相邻两级之间分支的总宽度保持不变（能量守恒），如下图：  数据准备 此处我们使用Metaphlan2 tutorial中的数据，来源于牙龈菌斑的宏基因组测序：
download.file(&#39;https://bitbucket.org/biobakery/biobakery/raw/tip/demos/biobakery_demos/data/metaphlan2/output/SRS014476-Supragingival_plaque_profile.txt&#39;, destfile = &#39;SRS014476-Supragingival_plaque_profile.txt&#39;) taxtab &lt;- read.table(&#39;SRS014476-Supragingival_plaque_profile.txt&#39;, sep=&#39;\t&#39;, stringsAsFactors=FALSE)  加载networkdD3包 该包使用D3.js实现数据的交互式可视化，具体一些例子参见这里：
library(networkD3)  数据预处理 函数sankeyNetwork主要需要两个data frame: Links和Nodes。Links主要有连接的起点（source）和终点（target），似乎此处的起点和终点只能是从0开始的数字。Nodes如果不指定ID，默认行是按照0开始的数字排列。
second_last &lt;- function(x) ifelse(length(x)&gt;1, x[length(x)-1], NA) taxtab &lt;- taxtab[-grep(&#39;unclassified&#39;, taxtab$V1), ] ## remove unlassified taxa tax_split &lt;- strsplit(taxtab$V1, &#39;\\|&#39;) ## split into different taxonomy levels target &lt;- sapply(tax_split, tail, n=1) ## target node mapping &lt;- data.frame(id=0:(length(target)-1), row.names=target) target_id &lt;- mapping$id source &lt;- sapply(tax_split, second_last) ## source node source_id &lt;- mapping[source,] value &lt;- taxtab$V2 ## width of flow links &lt;- data.</p>
  <footer class="post-footer">
    <a class="read-more" href="/post/2017-10-29-r-sankeyplot/">Read More →</a>
  </footer>
</article>

      
    
    
  </section>
  <footer class="site-footer">
  <p>© 2017 Chenhao&#39;s Blog</p>
  <p>Powered by <a href="https://gohugo.io/" target="_blank">Hugo</a> with theme <a href="https://github.com/laozhu/hugo-nuo" target="_blank">Nuo</a>.</p>
  
</footer>



<script src="/js/bundle.js"></script>


<script>
window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
ga('create', 'UA-108829179-1', 'auto');
ga('send', 'pageview');
</script>
<script async src='//www.google-analytics.com/analytics.js'></script>





  </body>
</html>
