---
title: "microbiomeViz--人生第一个R包"
author: "Chenhao Li （李陈浩）"
date: 2018-01-18
tags: ['R']
output: html_document
summary: "平日经常会分析shotgun宏基因组的数据，我们的[pipeline](https://github.com/gis-rpd/pipelines/tree/devel/metagenomics/shotgun-metagenomics)使用MetaPhlan，Kraken等profiler。一直觉得[GraPhlAn](https://bitbucket.org/nsegata/graphlan/wiki/Home)的图很好看，在Y叔的指导下准备基于ggtree来用R实现它。"
---



<section class="level3">
<h3>为啥写这个</h3>
<p>平日经常会分析shotgun宏基因组的数据，我们的<a href="https://github.com/gis-rpd/pipelines/tree/devel/metagenomics/shotgun-metagenomics">pipeline</a>使用MetaPhlan，Kraken等profiler。这种数据经常会产生一个表格，如下：</p>
<pre class="r"><code>download.file(&quot;https://bitbucket.org/biobakery/biobakery/raw/tip/demos/biobakery_demos/data/metaphlan2/output/SRS014459-Stool_profile.txt&quot;, &#39;~/Downloads/SRS014459-Stool_profile.txt&#39;)
knitr::kable(head(read.table(&#39;~/Downloads/SRS014459-Stool_profile.txt&#39;)))</code></pre>
<table>
<thead>
<tr class="header">
<th style="text-align: left;">V1</th>
<th style="text-align: right;">V2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">k__Bacteria</td>
<td style="text-align: right;">100.00000</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria|p__Firmicutes</td>
<td style="text-align: right;">64.91753</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria|p__Bacteroidetes</td>
<td style="text-align: right;">35.08247</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria|p__Firmicutes|c__Clostridia</td>
<td style="text-align: right;">64.91753</td>
</tr>
<tr class="odd">
<td style="text-align: left;">k__Bacteria|p__Bacteroidetes|c__Bacteroidia</td>
<td style="text-align: right;">35.08247</td>
</tr>
<tr class="even">
<td style="text-align: left;">k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales</td>
<td style="text-align: right;">64.91753</td>
</tr>
</tbody>
</table>
<p>第一列是分类信息注释，第二列是相对丰度（百分比）。在做这种图可视化方面，目前个人见过最强大的是<a href="https://bitbucket.org/nsegata/graphlan/wiki/Home">GraPhlAn</a>:</p>
<p><img src="https://bytebucket.org/nsegata/graphlan/wiki/GraPhlAn_Fig1.png" /></p>
<p>官网上相关的教程很详细，但是问题是，这个完全封闭的python程序，想要hack，还真的是挺难得。<a href="https://github.com/marbl/Krona/wiki">Krona</a>可能是另一个选择，但是同样还是会有同样的问题。最近发布的R包<a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005404">Metacoder</a>，画出的图个人真心不是很喜欢：</p>
<p><img src="http://journals.plos.org/ploscompbiol/article/figure/image?size=large&amp;id=info:doi/10.1371/journal.pcbi.1005404.g002" /></p>
<p>跟Y叔讨论了一下用ggtree实现像GraPhlAn那样图的可能性，得到了肯定的答复，于是开始自己造轮子。</p>
</section>
<section id="microbiomeviz" class="level3">
<h3>MicrobiomeViz–千里之行，始于足下</h3>
<p>其实可以写一个简单的函数，但是还是想做一个拓展性更强的东西，所以就有了这个包（其实还是只有一个函数：P）： <a href="https://github.com/lch14forever/microbiomeViz" class="uri">https://github.com/lch14forever/microbiomeViz</a></p>
</section>
<section id="metaphlanparser" class="level3">
<h3>MetaPhlan结果的parser</h3>
<section class="level4">
<h4>安装</h4>
<pre class="r"><code>devtools::install_github(&quot;lch14forever/microbiomeViz&quot;)</code></pre>
<pre><code>## Warning in strptime(x, fmt, tz = &quot;GMT&quot;): unknown timezone &#39;zone/tz/2017c.
## 1.0/zoneinfo/Asia/Singapore&#39;</code></pre>
<pre><code>## Skipping install of &#39;microbiomeViz&#39; from a github remote, the SHA1 (cf21e9f4) has not changed since last install.
##   Use `force = TRUE` to force installation</code></pre>
</section>
<section class="level4">
<h4>使用</h4>
<p>这个其实是基于之前做桑基图时写的那些代码，将MetaPhlan表格转化为了ape的phylo类：</p>
<pre class="r"><code>library(microbiomeViz)
tree &lt;- parseMetaphlanTSV(&#39;~/Downloads/SRS014459-Stool_profile.txt&#39;)</code></pre>
</section>
<section id="ggtree" class="level4">
<h4>ggtree的大戏</h4>
<p>画出主干</p>
<pre class="r"><code>library(ggtree)</code></pre>
<pre><code>## Warning: package &#39;ggtree&#39; was built under R version 3.4.3</code></pre>
<pre><code>## Loading required package: ggplot2</code></pre>
<pre><code>## Loading required package: treeio</code></pre>
<pre><code>## ggtree v1.10.2  For help: https://guangchuangyu.github.io/ggtree
## 
## If you use ggtree in published research, please cite:
## Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam. ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. Methods in Ecology and Evolution 2017, 8(1):28-36, doi:10.1111/2041-210X.12628</code></pre>
<pre class="r"><code>t &lt;- ggtree(tree, size=2, col=&#39;grey&#39;, layout = &#39;circular&#39;) + 
    geom_point(size=tree$nodeSize, shape=21, fill=&#39;white&#39;)
t</code></pre>
<p><img src="/post/2018-01-18-r-metagenomeViz_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>比如说我做了一个检验，发现Bacteroides属在疾病组显著增高，那么我们可以像GraPhlAn那样把它高亮出来：</p>
<pre class="r"><code>t &lt;- t + geom_hilight(node=22, fill = &#39;red&#39;, alpha = 0.3, extend = 1) +
   geom_text2(aes(subset=(node==22), label=label, vjust=-0.5))
t</code></pre>
<p><img src="/post/2018-01-18-r-metagenomeViz_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>其中B. ovatus也是在疾病组显著高：</p>
<pre class="r"><code>t &lt;- t + geom_hilight(node=7, fill = &#39;red&#39;, alpha = 0.3, extend = 1) +
   geom_text2(aes(subset=(node==7), label=label, vjust=1))
t</code></pre>
<p><img src="/post/2018-01-18-r-metagenomeViz_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>还想标注Clostridias在疾病组显著低：</p>
<pre class="r"><code>t &lt;- t + geom_hilight(node=13, fill = &#39;darkgreen&#39;, alpha = 0.3, extend = 1) +
   geom_text2(aes(subset=(node==13), label=label))
t</code></pre>
<p><img src="/post/2018-01-18-r-metagenomeViz_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
</section>
</section>
<section class="level3">
<h3>希望有更多的拓展</h3>
<p>肯定是不满足于只有这么个parser而已，其实有好多功能都想去实现：</p>
<ul>
<li>[ ] 多个MetaPhlan表，节点大小取平均</li>
<li>[ ] 根据统计检验结果自动添加高亮</li>
<li>[ ] 和heatmap结合</li>
<li>[ ] …</li>
</ul>
<p>这一切成为可能，都是因为有了ggtree这样一个灵活的框架。</p>
</section>
