<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.30" />
  <meta name="author" content="Chenhao Li">

  
  
  
  
    
      
    
  
  <meta name="description" content="Data downloading and preprocessing Now let’s try to generate a figure like this:
Sample processed by MetaPhlAn:
https://bitbucket.org/nsegata/metaphlan/wiki/profiled_samples.tar.bz2
Script to merge samples:
https://bitbucket.org/nsegata/metaphlan/raw/2f1b17a1f4e9775fe1ce42c8481279a5e69f291f/utils/merge_metaphlan_tables.py
Merge files into a single table
python merge_metaphlan_tables.py profiled_samples/*.txt &gt; profiled_samples/merged_abundance_table.txt  Now switch to R Load data and library df &lt;- read.table(&quot;~/Downloads/profiled_samples/merged_abundance_table.txt&quot;, head=TRUE, stringsAsFactors = FALSE) df &lt;- df[,-ncol(df)] ## Use row means as a proxy for node size dat &lt;- data.frame(V1=df[,1], V2=rowMeans(df[,-1]), stringsAsFactors = FALSE) library(microbiomeViz) ## Warning: replacing previous import &#39;ape::rotate&#39; by &#39;ggtree::rotate&#39; when ## loading &#39;microbiomeViz&#39;  Parse data and create a backbone tr &lt;- parseMetaphlanTSV(dat, node.">

  
  <link rel="alternate" hreflang="en-us" href="/post/2018-04-20-r-microbiomeviz_example/">

  


  

  
  
  <meta name="theme-color" content="#4caf50">
  
  
  
  
    
  
  
    
    
    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/atom-one-light.min.css">
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.1/css/academicons.min.css" integrity="sha512-NThgw3XKQ1absAahW6to7Ey42uycrVvfNfyjqcFNgCmOCQ5AR4AO0SiXrN+8ZtYeappp56lk1WtvjVmEa+VR6A==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.2.0/leaflet.css" integrity="sha512-M2wvCLH6DSRazYeZRIm1JnYyh22purTM+FDB5CsyxtQJYeKq83arPe5wgbNmcFXGqiSH2XR8dT/fJISVA1r/zQ==" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-108829179-1', 'auto');
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="Chenhao&#39;s Personal Page">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="Chenhao&#39;s Personal Page">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/2018-04-20-r-microbiomeviz_example/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@li_chenhao">
  <meta property="twitter:creator" content="@li_chenhao">
  
  <meta property="og:site_name" content="Chenhao&#39;s Personal Page">
  <meta property="og:url" content="/post/2018-04-20-r-microbiomeviz_example/">
  <meta property="og:title" content="Generating GraPhlAn-like microbiome visualization using ggtree and microbiomeViz | Chenhao&#39;s Personal Page">
  <meta property="og:description" content="Data downloading and preprocessing Now let’s try to generate a figure like this:
Sample processed by MetaPhlAn:
https://bitbucket.org/nsegata/metaphlan/wiki/profiled_samples.tar.bz2
Script to merge samples:
https://bitbucket.org/nsegata/metaphlan/raw/2f1b17a1f4e9775fe1ce42c8481279a5e69f291f/utils/merge_metaphlan_tables.py
Merge files into a single table
python merge_metaphlan_tables.py profiled_samples/*.txt &gt; profiled_samples/merged_abundance_table.txt  Now switch to R Load data and library df &lt;- read.table(&quot;~/Downloads/profiled_samples/merged_abundance_table.txt&quot;, head=TRUE, stringsAsFactors = FALSE) df &lt;- df[,-ncol(df)] ## Use row means as a proxy for node size dat &lt;- data.frame(V1=df[,1], V2=rowMeans(df[,-1]), stringsAsFactors = FALSE) library(microbiomeViz) ## Warning: replacing previous import &#39;ape::rotate&#39; by &#39;ggtree::rotate&#39; when ## loading &#39;microbiomeViz&#39;  Parse data and create a backbone tr &lt;- parseMetaphlanTSV(dat, node.">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2018-04-20T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2018-04-20T00:00:00&#43;00:00">
  

  

  <title>Generating GraPhlAn-like microbiome visualization using ggtree and microbiomeViz | Chenhao&#39;s Personal Page</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">Chenhao&#39;s Personal Page</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications_selected">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#teaching">
            
            <span>Teaching</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <div class="article-inner">
      <h1 itemprop="name">Generating GraPhlAn-like microbiome visualization using ggtree and microbiomeViz</h1>

      

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2018-04-20 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      Apr 20, 2018
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Chenhao Li">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    1 min read
  </span>
  

  
  

  

  
  

  

</div>


      <div class="article-style" itemprop="articleBody">
        <section id="data-downloading-and-preprocessing" class="level3">
<h3>Data downloading and preprocessing</h3>
<p>Now let’s try to generate a figure like this:</p>
<p><img src="https://bitbucket.org/nsegata/metaphlan/wiki/lefse_biomarkers_cladogram.png" /></p>
<p>Sample processed by MetaPhlAn:</p>
<p><a href="https://bitbucket.org/nsegata/metaphlan/wiki/profiled_samples.tar.bz2" class="uri">https://bitbucket.org/nsegata/metaphlan/wiki/profiled_samples.tar.bz2</a></p>
<p>Script to merge samples:</p>
<p><a href="https://bitbucket.org/nsegata/metaphlan/raw/2f1b17a1f4e9775fe1ce42c8481279a5e69f291f/utils/merge_metaphlan_tables.py" class="uri">https://bitbucket.org/nsegata/metaphlan/raw/2f1b17a1f4e9775fe1ce42c8481279a5e69f291f/utils/merge_metaphlan_tables.py</a></p>
<p>Merge files into a single table</p>
<pre class="sh"><code>python merge_metaphlan_tables.py profiled_samples/*.txt &gt; profiled_samples/merged_abundance_table.txt</code></pre>
</section>
<section id="now-switch-to-r" class="level3">
<h3>Now switch to R</h3>
<section id="load-data-and-library" class="level4">
<h4>Load data and library</h4>
<pre class="r"><code>df &lt;- read.table(&quot;~/Downloads/profiled_samples/merged_abundance_table.txt&quot;, head=TRUE, stringsAsFactors = FALSE)
df &lt;- df[,-ncol(df)]
## Use row means as a proxy for node size
dat &lt;- data.frame(V1=df[,1], V2=rowMeans(df[,-1]), stringsAsFactors = FALSE)

library(microbiomeViz)</code></pre>
<pre><code>## Warning: replacing previous import &#39;ape::rotate&#39; by &#39;ggtree::rotate&#39; when
## loading &#39;microbiomeViz&#39;</code></pre>
</section>
<section id="parse-data-and-create-a-backbone" class="level4">
<h4>Parse data and create a backbone</h4>
<pre class="r"><code>tr &lt;- parseMetaphlanTSV(dat, node.size.offset=2, node.size.scale=0.8)
p &lt;- tree.backbone(tr, size=0.5)
p</code></pre>
<p><img src="/post/2018-04-20-r-microbiomeviz_example_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
</section>
<section id="hard-coded-lefse-results" class="level4">
<h4>Hard-coded lefse results</h4>
<pre class="r"><code>lefse_lists = data.frame(node=c(&#39;s__Haemophilus_parainfluenzae&#39;,&#39;p__Proteobacteria&#39;,
                              &#39;f__Veillonellaceae&#39;,&#39;o__Selenomonadales&#39;,
                              &#39;c__Negativicutes&#39;, &#39;s__Streptococcus_parasanguinis&#39;,

                              &#39;p__Firmicutes&#39;,&#39;f__Streptococcaceae&#39;,
                              &#39;g__Streptococcus&#39;,&#39;o__Lactobacillales&#39;,
                              &#39;c__Bacilli&#39;,&#39;s__Streptococcus_mitis&#39;),
                         color=c(rep(&#39;darkgreen&#39;,6), rep(&#39;red&#39;,&#39;6&#39;)),
                         stringsAsFactors = FALSE
                         )</code></pre>
</section>
<section id="annotate-the-tree" class="level4">
<h4>Annotate the tree</h4>
<pre class="r"><code>p &lt;- clade.anno(p, lefse_lists, alpha=0.3)
p</code></pre>
<p><img src="/post/2018-04-20-r-microbiomeviz_example_files/figure-html/unnamed-chunk-5-1.png" width="960" /></p>
</section>
</section>

      </div>

      


<div class="article-tags">
  
  <a class="btn btn-primary btn-outline" href="/tags/r">R</a>
  
  <a class="btn btn-primary btn-outline" href="/tags/microbiomeviz">microbiomeViz</a>
  
</div>



    </div>
  </div>

</article>



<div class="article-container article-widget">
  <div class="hr-light"></div>
  <h3>Related</h3>
  <ul>
    
    <li><a href="/post/2018-02-01-r-metagenomeviz-italictaxon/">microbiomeViz--斜体显示taxon name</a></li>
    
    <li><a href="/post/2018-01-18-r-metagenomeviz/">microbiomeViz--人生第一个R包</a></li>
    
    <li><a href="/post/2018-03-29-r-base2grob_igraph/">Y叔的新包base2grob</a></li>
    
    <li><a href="/post/2018-01-12-r-compositional/">成分之咒与ReBoot算法</a></li>
    
    <li><a href="/post/2018-01-09-r-ordellipsecode/">ggplot扩展学习笔记--逐行解析Y叔的&#34;geom_ord_ellipse.R&#34;</a></li>
    
  </ul>
</div>




<div class="article-container">
  
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "lichenhao" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>


</div>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      © This site was created with R BlogDown and HuGo by Chenhao Li. &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.2.0/leaflet.js" integrity="sha512-lInM/apFSqyy1o6s89K4iQUKg6ppXEgsVxT35HbzUupEVRh2Eu9Wdl4tHj7dZO0s1uvplcYGmt3498TtHq+log==" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/R.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/sh.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

