<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.30.2" />
  <meta name="author" content="Chenhao Li">

  
  
  
  
    
      
    
  
  <meta name="description" content="微生物组研究走向绝对定量 Jeroen Raes研究组上周在Nature发表文章，使用Flow cytometry估算粪便中的微生物细胞数量，配合16S测序算出的相对丰度，估算出OTU的绝对数量（absolute abundance）。文章很多结论都印证了直接应用相对丰度进行分析时，我们所看到的很多现象是成分数据性质的假象（可以想象，由于相对丰度在每一个样本中相加为1或100，一个OTU相对丰度增加必将引起其他OTU相对丰度减少，所谓的compositional bias）。特别针对于计算两个OTU的相关系数，当OTU分布不均匀时，很容易看到负相关的OTU–而这仅仅是因为它们受到相加为常数的限制而已。另一个典型例子就是主成分分析（PCA），PCA意在保持欧式距离不变的情况下对数据进行变换，但是相对丰度其实不在欧式空间中（可以参考:J. Aitchison, The Statistical Analysis of Compositional Data, 1986.），这就是为什么在微生物组的研究中更多采用生态学的距离（如，Bray-Curtis distance）来计算\(\beta\) -diversity，然后进行基于距离矩阵的分析（PCoA）。
似乎文章的take home message很简单了，微生物组的研究，我们应该使用类似的方法进行绝对定量。可是仔细想一想，文章指出粪便微生物总量的个体差异可以达到10倍之多，这样大的差异，如果某种微生物在个体之间差异很小，转换成绝对数量之后，个体间的差异将受制于微生物总量。
 使用Flow cytometry测定的肠道微生物总量的variation有多大？ 下载Nature文章中的Supplementary Table，这里并不不需要购买文章阅读权限，其中表6是Flow cytometry的数据。
加载R包，ggplot2主题 library(tidyverse) ## ── Attaching packages ─────────────────────────────────────────────────────────── tidyverse 1.2.1 ── ## ✔ ggplot2 2.2.1 ✔ purrr 0.2.4 ## ✔ tibble 1.3.4 ✔ dplyr 0.7.4 ## ✔ tidyr 0.7.2 ✔ stringr 1.2.0 ## ✔ readr 1.1.1 ✔ forcats 0.2.0 ## ── Conflicts ────────────────────────────────────────────────────────────── tidyverse_conflicts() ── ## ✖ dplyr::filter() masks stats::filter() ## ✖ dplyr::lag() masks stats::lag() library(data.">

  
  <link rel="alternate" hreflang="en-us" href="/post/2017-11-18-r-absolute/">

  


  

  
  
  <meta name="theme-color" content="#4caf50">
  
  
  
  
    
  
  
    
    
    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/solarized-dark.min.css">
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.1/css/academicons.min.css" integrity="sha512-NThgw3XKQ1absAahW6to7Ey42uycrVvfNfyjqcFNgCmOCQ5AR4AO0SiXrN+8ZtYeappp56lk1WtvjVmEa+VR6A==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.2.0/leaflet.css" integrity="sha512-M2wvCLH6DSRazYeZRIm1JnYyh22purTM+FDB5CsyxtQJYeKq83arPe5wgbNmcFXGqiSH2XR8dT/fJISVA1r/zQ==" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  

  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-108829179-1', 'auto');
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="//www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="Chenhao&#39;s Blog">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="Chenhao&#39;s Blog">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/2017-11-18-r-absolute/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@li_chenhao">
  <meta property="twitter:creator" content="@li_chenhao">
  
  <meta property="og:site_name" content="Chenhao&#39;s Blog">
  <meta property="og:url" content="/post/2017-11-18-r-absolute/">
  <meta property="og:title" content="有了绝对定量就完了？ | Chenhao&#39;s Blog">
  <meta property="og:description" content="微生物组研究走向绝对定量 Jeroen Raes研究组上周在Nature发表文章，使用Flow cytometry估算粪便中的微生物细胞数量，配合16S测序算出的相对丰度，估算出OTU的绝对数量（absolute abundance）。文章很多结论都印证了直接应用相对丰度进行分析时，我们所看到的很多现象是成分数据性质的假象（可以想象，由于相对丰度在每一个样本中相加为1或100，一个OTU相对丰度增加必将引起其他OTU相对丰度减少，所谓的compositional bias）。特别针对于计算两个OTU的相关系数，当OTU分布不均匀时，很容易看到负相关的OTU–而这仅仅是因为它们受到相加为常数的限制而已。另一个典型例子就是主成分分析（PCA），PCA意在保持欧式距离不变的情况下对数据进行变换，但是相对丰度其实不在欧式空间中（可以参考:J. Aitchison, The Statistical Analysis of Compositional Data, 1986.），这就是为什么在微生物组的研究中更多采用生态学的距离（如，Bray-Curtis distance）来计算\(\beta\) -diversity，然后进行基于距离矩阵的分析（PCoA）。
似乎文章的take home message很简单了，微生物组的研究，我们应该使用类似的方法进行绝对定量。可是仔细想一想，文章指出粪便微生物总量的个体差异可以达到10倍之多，这样大的差异，如果某种微生物在个体之间差异很小，转换成绝对数量之后，个体间的差异将受制于微生物总量。
 使用Flow cytometry测定的肠道微生物总量的variation有多大？ 下载Nature文章中的Supplementary Table，这里并不不需要购买文章阅读权限，其中表6是Flow cytometry的数据。
加载R包，ggplot2主题 library(tidyverse) ## ── Attaching packages ─────────────────────────────────────────────────────────── tidyverse 1.2.1 ── ## ✔ ggplot2 2.2.1 ✔ purrr 0.2.4 ## ✔ tibble 1.3.4 ✔ dplyr 0.7.4 ## ✔ tidyr 0.7.2 ✔ stringr 1.2.0 ## ✔ readr 1.1.1 ✔ forcats 0.2.0 ## ── Conflicts ────────────────────────────────────────────────────────────── tidyverse_conflicts() ── ## ✖ dplyr::filter() masks stats::filter() ## ✖ dplyr::lag() masks stats::lag() library(data.">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2017-11-28T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2017-11-28T00:00:00&#43;00:00">
  

  

  <title>有了绝对定量就完了？ | Chenhao&#39;s Blog</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">Chenhao&#39;s Blog</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#publications_selected">
            
            <span>Publications</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#posts">
            
            <span>Posts</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#teaching">
            
            <span>Teaching</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <div class="article-inner">
      <h1 itemprop="name">有了绝对定量就完了？</h1>

      

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2017-11-28 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      Nov 28, 2017
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Chenhao Li">
  </span>

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    5 min read
  </span>
  

  
  

  
  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fa fa-folder"></i>
    
    <a href="/categories/%e8%af%bb%e6%96%87%e7%8c%ae">读文献</a
    >
    
  </span>
  
  

  
  

  

</div>


      <div class="article-style" itemprop="articleBody">
        <section class="level2">
<h2>微生物组研究走向绝对定量</h2>
<p>Jeroen Raes研究组上周在Nature发表<a href="https://www.nature.com/articles/nature24460">文章</a>，使用Flow cytometry估算粪便中的微生物细胞数量，配合16S测序算出的相对丰度，估算出OTU的绝对数量（absolute abundance）。文章很多结论都印证了直接应用相对丰度进行分析时，我们所看到的很多现象是成分数据性质的假象（可以想象，由于相对丰度在每一个样本中相加为1或100，一个OTU相对丰度增加必将引起其他OTU相对丰度减少，所谓的compositional bias）。特别针对于计算两个OTU的相关系数，当OTU分布不均匀时，很容易看到负相关的OTU–而这仅仅是因为它们受到相加为常数的限制而已。另一个典型例子就是主成分分析（PCA），PCA意在保持欧式距离不变的情况下对数据进行变换，但是相对丰度其实不在欧式空间中（可以参考:J. Aitchison, The Statistical Analysis of Compositional Data, 1986.），这就是为什么在微生物组的研究中更多采用生态学的距离（如，Bray-Curtis distance）来计算<span class="math inline">\(\beta\)</span> -diversity，然后进行基于距离矩阵的分析（PCoA）。</p>
<p>似乎文章的take home message很简单了，微生物组的研究，我们应该使用类似的方法进行绝对定量。可是仔细想一想，文章指出粪便微生物总量的个体差异可以达到10倍之多，这样大的差异，如果某种微生物在个体之间差异很小，转换成绝对数量之后，个体间的差异将受制于微生物总量。</p>
</section>
<section id="flow-cytometryvariation" class="level2">
<h2>使用Flow cytometry测定的肠道微生物总量的variation有多大？</h2>
<p>下载Nature文章中的<a href="https://images.nature.com/original/nature-assets/nature/journal/vaop/ncurrent/extref/nature24460-s2.xlsx">Supplementary Table</a>，这里并不不需要购买文章阅读权限，其中表6是Flow cytometry的数据。</p>
<section id="rggplot2" class="level3">
<h3>加载R包，ggplot2主题</h3>
<pre class="r"><code>library(tidyverse)</code></pre>
<pre><code>## ── Attaching packages ─────────────────────────────────────────────────────────── tidyverse 1.2.1 ──</code></pre>
<pre><code>## ✔ ggplot2 2.2.1     ✔ purrr   0.2.4
## ✔ tibble  1.3.4     ✔ dplyr   0.7.4
## ✔ tidyr   0.7.2     ✔ stringr 1.2.0
## ✔ readr   1.1.1     ✔ forcats 0.2.0</code></pre>
<pre><code>## ── Conflicts ────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
## ✖ dplyr::filter() masks stats::filter()
## ✖ dplyr::lag()    masks stats::lag()</code></pre>
<pre class="r"><code>library(data.table)</code></pre>
<pre><code>## 
## Attaching package: &#39;data.table&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:dplyr&#39;:
## 
##     between, first, last</code></pre>
<pre><code>## The following object is masked from &#39;package:purrr&#39;:
## 
##     transpose</code></pre>
<pre class="r"><code>library(cowplot)</code></pre>
<pre><code>## 
## Attaching package: &#39;cowplot&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:ggplot2&#39;:
## 
##     ggsave</code></pre>
<pre class="r"><code>library(readxl)

figtheme &lt;- theme_bw() + 
  theme(text = element_text(size=10,face=&#39;bold&#39;),panel.border  = element_rect(colour = &quot;black&quot;,size=2),
        axis.title.y=element_text(margin=margin(0,15,0,0)),axis.title.x=element_text(margin=margin(15,0,0,0)),
        plot.margin = unit(c(1,1,1,1), &quot;cm&quot;),
        plot.title = element_text(margin=margin(0,0,15,0)))
theme_set(figtheme)</code></pre>
</section>
<section id="6" class="level3">
<h3>读取表6，并预处理</h3>
<pre class="r"><code>cell_counts_dat &lt;- read_excel(&#39;~/nature24460-s2.xlsx&#39;, sheet=6) %&gt;%
  mutate(Cohort=gsub(&#39;[0-9]+&#39;,&#39;&#39;,Individual)) %&gt;%
  mutate(Cell_count_avg=`Average cell count (per gram of frozen feces)`)</code></pre>
<pre><code>## Warning in strptime(x, format, tz = tz): unknown timezone &#39;zone/tz/2017c.
## 1.0/zoneinfo/Asia/Singapore&#39;</code></pre>
<pre class="r"><code>cell_counts_dat</code></pre>
<pre><code>## # A tibble: 321 x 10
##    Individual Cohort   Day `Health status`
##         &lt;chr&gt;  &lt;chr&gt; &lt;dbl&gt;           &lt;chr&gt;
##  1       SC01     SC     1         Healthy
##  2       SC02     SC     1         Healthy
##  3       SC03     SC     1         Healthy
##  4       SC04     SC     1         Healthy
##  5       SC05     SC     1         Healthy
##  6       SC06     SC     1         Healthy
##  7       SC07     SC     1         Healthy
##  8       SC08     SC     1         Healthy
##  9       SC09     SC     1         Healthy
## 10       SC10     SC     1         Healthy
## # ... with 311 more rows, and 6 more variables: `Average cell count (per
## #   gram of fresh feces)` &lt;chr&gt;, `STDEV cell count (per gram of fresh
## #   feces)` &lt;chr&gt;, `Average cell count (per gram of frozen feces)` &lt;dbl&gt;,
## #   `STDEV cell count (per gram of frozen feces)` &lt;chr&gt;, Enterotype &lt;chr&gt;,
## #   Cell_count_avg &lt;dbl&gt;</code></pre>
</section>
<section class="level3">
<h3>微生物总量差异</h3>
<ul>
<li>个体差异</li>
</ul>
<pre class="r"><code>aggregate(data=cell_counts_dat, Cell_count_avg~Cohort, function(x) c(Mean=mean(x), SD=sd(x), CV=sd(x)/mean(x)) )</code></pre>
<pre><code>##   Cohort Cell_count_avg.Mean Cell_count_avg.SD Cell_count_avg.CV
## 1     DC        9.092967e+10      6.344747e+10      6.977642e-01
## 2     LC        1.249278e+11      5.406697e+10      4.327858e-01
## 3     SC        1.528339e+11      7.018824e+10      4.592451e-01
## 4     VC        1.447904e+11      7.630896e+10      5.270304e-01</code></pre>
<pre class="r"><code>ggplot(cell_counts_dat, aes(x=Cohort, y=Cell_count_avg)) + geom_boxplot()</code></pre>
<p><img src="/post/2017-11-18-r-absolute_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<ul>
<li>个体内差异</li>
</ul>
<pre class="r"><code>aggregate(data=cell_counts_dat[cell_counts_dat$Cohort==&#39;LC&#39;,], Cell_count_avg~Individual, function(x) c(Mean=mean(x), SD=sd(x), CV=sd(x)/mean(x)) )</code></pre>
<pre><code>##    Individual Cell_count_avg.Mean Cell_count_avg.SD Cell_count_avg.CV
## 1        LC01        1.121069e+11      3.062958e+10      2.732175e-01
## 2        LC02        1.057725e+11      1.457955e+10      1.378387e-01
## 3        LC03        6.998669e+10      3.207140e+10      4.582500e-01
## 4        LC04        6.433548e+10      1.478855e+10      2.298661e-01
## 5        LC05        2.198245e+11      2.726579e+10      1.240343e-01
## 6        LC06        2.054927e+11      4.951278e+10      2.409466e-01
## 7        LC07        1.197242e+11      3.438372e+10      2.871911e-01
## 8        LC08        1.162890e+11      3.286685e+10      2.826307e-01
## 9        LC09        1.501435e+11      6.018554e+10      4.008535e-01
## 10       LC10        1.790086e+11      8.293900e+10      4.633241e-01
## 11       LC11        1.565348e+11      4.038210e+10      2.579752e-01
## 12       LC12        1.247153e+11      2.494971e+10      2.000533e-01
## 13       LC13        9.804859e+10      2.625298e+10      2.677548e-01
## 14       LC14        1.037225e+11      4.329933e+10      4.174534e-01
## 15       LC15        1.592359e+11      2.337419e+10      1.467897e-01
## 16       LC16        1.171450e+11      6.207212e+10      5.298741e-01
## 17       LC17        1.215423e+11      3.782074e+10      3.111736e-01
## 18       LC18        1.042144e+11      4.236728e+10      4.065395e-01
## 19       LC19        7.013542e+10      2.326036e+10      3.316492e-01
## 20       LC20        1.465100e+11      2.763866e+10      1.886470e-01</code></pre>
<pre class="r"><code>ggplot(cell_counts_dat[cell_counts_dat$Cohort==&#39;LC&#39;,], aes(x=Day, y=Cell_count_avg)) + geom_line() + facet_wrap(~Individual)</code></pre>
<p><img src="/post/2017-11-18-r-absolute_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>容易看出，疾病组（DC）的微生物总量最低，其他几个健康组微生物总量差异不大，而个人的微生物总量在一周内的变化要小一些（CV~30%），但其实比起个体间差异不是小很多。</p>
</section>
</section>
<section id="variation" class="level2">
<h2>人的肠道微生物相对丰度的variation有多大？</h2>
<section id="hmp" class="level3">
<h3>下载人类微生物组项目（HMP）数据</h3>
<p>我们使用HMPv35的16S的<a href="ftp://public-ftp.ihmpdcc.org/HMQCP/otu_table_psn_v35.txt.gz">OTU table</a>和元数据表(<a href="ftp://public-ftp.ihmpdcc.org/HMQCP/v35_map_uniquebyPSN.txt.bz2">mapping file</a>)。下载之后解压。</p>
</section>
<section id="0otu" class="level3">
<h3>清洗数据，保留肠道数据并去掉整行为0的OTU</h3>
<pre class="r"><code>metadata &lt;- fread(&#39;~/v35_map_uniquebyPSN.txt&#39;)
stoolIDs &lt;- as.character((metadata %&gt;% filter(HMPbodysubsite==&#39;Stool&#39;))[,1])

otutab &lt;- fread(&#39;~/otu_table_psn_v35.txt&#39;,head=TRUE, select=stoolIDs,sep=&#39;\t&#39;)</code></pre>
<pre><code>## Warning in fread(&quot;~/otu_table_psn_v35.txt&quot;, head = TRUE, select =
## stoolIDs, : Starting data input on line 2 and discarding line 1 because it
## has too few or too many items to be column names or data: # QIIME v1.3.0-
## dev OTU table</code></pre>
<pre><code>## Warning in fread(&quot;~/otu_table_psn_v35.txt&quot;, head = TRUE, select =
## stoolIDs, : Column name &#39;700107040&#39; not found in column name header (case
## sensitive), skipping.</code></pre>
<pre><code>## Warning in fread(&quot;~/otu_table_psn_v35.txt&quot;, head = TRUE, select =
## stoolIDs, : Column name &#39;700114489&#39; not found in column name header (case
## sensitive), skipping.</code></pre>
<pre><code>## 
Read 44.1% of 45383 rows
Read 45383 rows and 319 (of 4790) columns from 0.409 GB file in 00:00:04</code></pre>
<pre class="r"><code>## keep samples with counts &gt; 2000
otutab &lt;- otutab %&gt;% select(which(colSums(otutab)&gt; 2000 ))
## we down sample (rarefy) the samples to have 2000 counts in total
otutab_rare &lt;- apply(otutab, 2, function(x) rmultinom(1,size=2000, prob=x))
## remove OTUs not present at all and normalize to proportions
otutab_rel &lt;- apply(otutab_rare[rowSums(otutab_rare)!=0,], 2, function(x)x/sum(x))</code></pre>
</section>
<section id="sd-cv-mean" class="level3">
<h3>SD, CV, Mean</h3>
<pre class="r"><code>SD &lt;- apply(otutab_rel,1,sd)
Mean &lt;- apply(otutab_rel, 1, mean)
tmp &lt;- data.frame(CV=SD/Mean, Mean, SD)
g1 &lt;- ggplot(tmp, aes(y=SD, x=Mean)) + 
  geom_point(alpha=0.1) + 
  scale_x_log10()  + scale_y_log10() + 
  annotation_logticks(sides = &#39;lb&#39;)
g2 &lt;- ggplot(tmp, aes(y=CV, x=Mean)) + 
  geom_point(alpha=0.1) + 
  scale_x_log10()   + 
  annotation_logticks(sides = &#39;b&#39;)
plot_grid(g1, g2, nrow=1)</code></pre>
<p><img src="/post/2017-11-18-r-absolute_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>从CV来看，HMP的数据相对丰度其实在个体间差距比较大（特别是低丰度的OTU，可以想象这是跟0的数量多有关），所以可能大多数情况下，我们不会受到总量变化（CV~50%）的影响。<strong>Note: </strong>还是需要注意在我们的研究数据中会不会</p>
</section>
</section>
<section id="a-thought-experiment" class="level2">
<h2>绝对定量的分析，A thought experiment</h2>
<p>正态分布的随机变量（<span class="math inline">\(M\)</span>，<span class="math inline">\(X\)</span>）分别代表微生物总量（<span class="math inline">\(M\sim|Normal(10^{11},CV_m\times10^{11}|)\)</span>），微生物<span class="math inline">\(x\)</span>的相对丰度（<span class="math inline">\(X\sim| Normal(10^{-1},CV_x \times10^{-1})|\)</span>），那么绝对数量<span class="math inline">\(MX\)</span>会与<span class="math inline">\(M\)</span>相关吗？</p>
<pre class="r"><code>M &lt;- abs(rnorm(1000, 1e11, 0.5e11)) ## CV=0.5
X &lt;- sapply(c(0.1,0.5,1,5,10), function(x)abs(rnorm(1000, 0.1, x*0.1)))   
colnames(X) &lt;- c(&#39;X_cv0.1&#39;,&#39;X_cv0.5&#39;,&#39;X_cv1&#39;, &#39;X_cv5&#39;, &#39;X_cv10&#39;)
dat &lt;- melt(X)%&gt;% mutate(MX=value*M) %&gt;% mutate(M=rep(M,5))

dat %&gt;% group_by(Var2) %&gt;% summarise(cor(MX, M, method=&#39;spearman&#39;))</code></pre>
<pre><code>## # A tibble: 5 x 2
##      Var2 `cor(MX, M, method = &quot;spearman&quot;)`
##    &lt;fctr&gt;                             &lt;dbl&gt;
## 1 X_cv0.1                         0.9745780
## 2 X_cv0.5                         0.6241846
## 3   X_cv1                         0.4850720
## 4   X_cv5                         0.5014045
## 5  X_cv10                         0.4692217</code></pre>
<pre class="r"><code>ggplot(dat,aes(x=M,y=MX)) + 
  geom_point(alpha=0.4) + 
  facet_wrap(~Var2, nrow=2, scale=&#39;free_y&#39;)</code></pre>
<p><img src="/post/2017-11-18-r-absolute_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>所以不难看出来微生物的绝对数量与微生物总量相关，特别是当相对丰度差异不大的时候。So what？记得correlation有传递性吗？所以如果有两个不相关的微生物，它们的绝对数量会不会因为总量而变得相关呢？我们根据<span class="math inline">\(X\)</span>的分布生成另一个独立的随机变量<span class="math inline">\(Y\)</span>，让我们来看下他们的correlation是怎样的。</p>
<pre class="r"><code>Y &lt;- sapply(c(0.1,0.5,1,5,10), function(x)abs(rnorm(1000, 0.1, x*0.1))) 
colnames(Y) &lt;- c(&#39;Y_cv0.1&#39;,&#39;Y_cv0.5&#39;,&#39;Y_cv1&#39;, &#39;Y_cv5&#39;, &#39;Y_cv10&#39;)
dat &lt;- dat %&gt;% mutate(Y=melt(Y)$value) %&gt;% mutate(MY=M*Y)</code></pre>
<ul>
<li>毫无关联的<span class="math inline">\(X\)</span>,<span class="math inline">\(Y\)</span></li>
</ul>
<pre class="r"><code>dat %&gt;% group_by(Var2) %&gt;% summarise(cor(value, Y, method=&#39;spearman&#39;))</code></pre>
<pre><code>## # A tibble: 5 x 2
##      Var2 `cor(value, Y, method = &quot;spearman&quot;)`
##    &lt;fctr&gt;                                &lt;dbl&gt;
## 1 X_cv0.1                          0.039534460
## 2 X_cv0.5                          0.025382017
## 3   X_cv1                          0.039640060
## 4   X_cv5                          0.015882484
## 5  X_cv10                          0.002462222</code></pre>
<ul>
<li>转化成绝对数量后</li>
</ul>
<pre class="r"><code>dat %&gt;% group_by(Var2) %&gt;% 
  summarise(Spearman_rho=cor(MX, MY, method=&#39;spearman&#39;),
            p_value=cor.test(MX, MY, method=&#39;spearman&#39;)$p.value)</code></pre>
<pre><code>## # A tibble: 5 x 3
##      Var2 Spearman_rho      p_value
##    &lt;fctr&gt;        &lt;dbl&gt;        &lt;dbl&gt;
## 1 X_cv0.1    0.9509044 0.000000e+00
## 2 X_cv0.5    0.4455341 0.000000e+00
## 3   X_cv1    0.3194627 0.000000e+00
## 4   X_cv5    0.2866312 2.842296e-20
## 5  X_cv10    0.2930657 3.241526e-21</code></pre>
<ul>
<li>散点图</li>
</ul>
<pre class="r"><code>ggplot(dat,aes(x=MX,y=MY)) + 
  geom_point(alpha=0.4) + 
  facet_wrap(~Var2, nrow=2, scale=&#39;free&#39;)</code></pre>
<p><img src="/post/2017-11-18-r-absolute_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
</section>
<section class="level2">
<h2>相关性？为什么相关？</h2>
<p>所以在以上的实验当中，不难看出即使不相关的两个OTU，因为个体差异不大，转化为绝对数量之后很有可能呈现相关性，而这样的相关性跟生态意义毫无关系，只是因为latent variable微生物总量而已（Nature文章的Figure 3）。</p>
<p>延展开来，当你发现某些微生物的绝对数量和某种表征相关的时候，也需要考虑一下，它们是否只是因为微生物总量的不同呢（Nature文章的Figure 4）？</p>
</section>

      </div>

      


<div class="article-tags">
  
  <a class="btn btn-primary btn-outline" href="/tags/r">R</a>
  
  <a class="btn btn-primary btn-outline" href="/tags/metagenomics">Metagenomics</a>
  
  <a class="btn btn-primary btn-outline" href="/tags/correlation-network">Correlation network</a>
  
</div>



    </div>
  </div>

</article>



<div class="article-container article-widget">
  <div class="hr-light"></div>
  <h3>Related</h3>
  <ul>
    
    <li><a href="/post/2017-11-15-r-glv/">Generalized Lotka-Volterra model</a></li>
    
    <li><a href="/post/2017-11-07-r-curatedmetagenomicdata_retrievedata/">从curatedMetagenoimcData提取健康样本的微生物组</a></li>
    
    <li><a href="/post/2017-11-02-r-sankeyplot2/">桑基图（Sankey diagram）II</a></li>
    
    <li><a href="/post/2017-10-30-r-curatedmetagenomicdata/">超大宏基因组数据集CuratedMetagenomicData</a></li>
    
    <li><a href="/post/2017-10-29-r-sankeyplot/">桑基图（Sankey diagram）</a></li>
    
  </ul>
</div>




<div class="article-container">
  
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "lichenhao" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>


</div>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      © This site was created with R BlogDown and HuGo by Chenhao Li. &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    

    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.2.0/leaflet.js" integrity="sha512-lInM/apFSqyy1o6s89K4iQUKg6ppXEgsVxT35HbzUupEVRh2Eu9Wdl4tHj7dZO0s1uvplcYGmt3498TtHq+log==" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/R.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/sh.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

